### ISSUE TITLE
How to enable clang-analyzer-unix.MismatchedDeallocator

### ISSUE THREAD

**Issue by timxx** [clangd/clangd#905]
Even add `clang-analyzer-unix.MismatchedDeallocator` check to Diagnostics ClangTidy section, no effects with clangd, but it works with clang-tidy checks.

Seems only *-Wmismatched-new-delete* diagnostics works.

**System information**

Output of `clangd --version`:
13.0.0/12.0.1

Editor/LSP plugin:
VS Code/clangd(v0.1.13)
Operating system:
Windows/ArchLinux

**Comment by Trass3r** [comment]
Guess you'll have to leave the static analyzer to the CI.

**Comment by HighCommander4** [comment]
One could also consider writing a dedicated clang static analyzer plugin for vscode, which runs the static analysis on demand (not in a long-running process) and visualizes its diagnostics in the Problems view.

**Comment by g-berthiaume** [comment]
**Addedum:**  
To my understanding, https://github.com/Ericsson/codechecker is a tool that will run `scan-build` automatically after the file has been saved. It also have a [vscode extension](https://marketplace.visualstudio.com/items?itemName=codechecker.vscode-codechecker).

Not sure about the limitations.

**Comment by Trass3r** [comment]
> One could also consider writing a dedicated clang static analyzer plugin

I'd love to have some real UI Integration with visualized/navigable chains as shown in the screenshots: https://clang-analyzer.llvm.org/

I guess one could write an extension that reads the compile_commands.json and executes the commands with `--analyze` and then somehow shows the results in the UI.
Maybe just the html version, or by generating SARIF and using https://marketplace.visualstudio.com/items?itemName=MS-SarifVSCode.sarif-viewer.

**Comment by studgeek** [comment]
Is there a way to identify which clang-tidy checks are not included? https://clang.llvm.org/extra/clang-tidy/checks/list.html doesn't seem to indicate which checks depend on the clang static analyzer.

**Comment by ByunghoonKim** [comment]
I believe `clangd` filters out without notification not only `clang-anlyzer-*` but also some other time consuming checks by default. For example, `clangd` in my `neovim` editor ignores `misc-const-correctness` from my `.clang-tidy` configuration
```C++
// snippet from https://clang.llvm.org/extra/clang-tidy/checks/misc/const-correctness.html
int foo()
{
    // Declare a variable, which is not ``const`` ...
    int i = 42;
    // but use it as read-only. This means that `i` can be declared ``const``.
    int result = i * i;

    return result;
}
```
, while command line `clang-tidy` complains
```
/Users/.../asdf.cxx:4:5: warning: variable 'i' of type 'int' can 
be declared 'const' [misc-const-correctness]
    4 |     int i = 42;
      |     ^
      |         const 
/Users/.../asdf.cxx:6:5: warning: variable 'result' of type 'int'
 can be declared 'const' [misc-const-correctness]
    6 |     int result = i * i;
      |     ^
      |         const 
```
@studgeek, I wish `clangd` has an option to print out all the checks it has decided to ignore silently.

See also https://clangd.llvm.org/config#fastcheckfilter.

> FastCheckFilter[#](https://clangd.llvm.org/config#fastcheckfilter)
> 
> Whether to run clang-tidy checks that may slow down clangd.
> 
> Valid values are:
> 
> - Strict: Run only checks measured to be fast. This excludes recently-added checks that we have not timed yet. This is the default.
> - Loose: Run checks unless they are known to be slow.
> - None: Run checks regardless of their speed.

**Comment by HighCommander4** [comment]
That's right, there are some checkers that clangd disables for reasons unrelated to the static analyzer:

 * Checkers disabled for performance reasons, as discussed in the previous comment. This can be overridden with the `FastCheckFilter` setting. A list of "known to be fast" and "known to be slow" checkers can be found in [this file](https://searchfox.org/llvm/rev/73e14de207a3aa0fa071fa56756e8e316edf5227/clang-tools-extra/clangd/TidyFastChecks.inc).
 * A few checkers disabled unconditionally because they're crashy or otherwise incompatible with clangd. These are listed [here](https://searchfox.org/llvm/rev/73e14de207a3aa0fa071fa56756e8e316edf5227/clang-tools-extra/clangd/TidyProvider.cpp#208-225).

**Comment by bernardassan** [comment]
The only reliable way I have found to use clang static analyzer in my editor is by integrating it into [efm-langserver](https://github.com/bernardassan/awesome-helix/blob/833ddc0d28f53dbb9545ea7dbac72080e6bb0cdf/efm-langserver/config.yaml#L88) so I get diagnostics on save. 

Combined with a [.clangd](https://github.com/bernardassan/czalloc/blob/2c5744132d758834f7cdb03a3400f49278443636/.clangd) in each specific projects, I get the best of both worlds but it would be nice if this issue is solved so that such workarounds wouldn't be necessary

**Comment by dabrahams** [comment]
I have used CodeChecker and its extension; they work.  The issues are:

1. At least on my mac, the easy install methods [don't work](https://github.com/Ericsson/codechecker/issues/4676).  I had to build CodeChecker from source. The macOS guide is [here](https://codechecker.readthedocs.io/en/latest/install_macosx/), but don't disable SIP because
2. the `CodeChecker log` command which supposedly needs SIP disabled [doesn't work](https://github.com/Ericsson/codechecker/issues/4678), even when you disable SIP.  If you need that functionality you should use [Bear](https://github.com/rizsotto/Bear) instead.
